utils/network.go

```diff
@@ -177,10 +200,93 @@ func (n Network) CoreContractAddress() (common.Address, error) {
 	return address, nil
 }
 
-func (n Network) ChainID() *felt.Felt {
+func (n NetworkKnown) ChainID() *felt.Felt {
 	return new(felt.Felt).SetBytes([]byte(n.ChainIDString()))
 }
 
-func (n Network) ProtocolID() protocol.ID {
+func (n NetworkKnown) ProtocolID() protocol.ID {
 	return protocol.ID(fmt.Sprintf("/starknet/%s", n))
 }
+
+type NetworkCustom struct {
+	FeederURLVal           string          `yaml:"feeder_url" validate:"required"`
+	GatewayURLVal          string          `yaml:"gateway_url" validate:"required"`
+	ChainIDVal             string          `yaml:"chain_id" validate:"required"`
+	L1ChainIDVal           *big.Int        `yaml:"l1_chain_id" validate:"required"`
+	ProtocolIDVal          int             `yaml:"protocol_id" validate:"required"`
+	CoreContractAddressVal *common.Address `yaml:"core_contract_address" validate:"required"`
+}
+
+func (cn *NetworkCustom) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type Alias NetworkCustom
+	aux := &struct {
+		CoreContractAddress string `yaml:"core_contract_address"`
+		*Alias
+	}{
+		Alias: (*Alias)(cn),
+	}
+	if err := unmarshal(aux); err != nil {
+		return err
+	}
+	if !common.IsHexAddress(aux.CoreContractAddress) {
+		return ErrInvalidCoreContractAddress
+	}
+	coreContractAddress := common.HexToAddress(aux.CoreContractAddress)
+	cn.CoreContractAddressVal = &coreContractAddress
+	return nil
+}
+
+func (cn NetworkCustom) String() string {
+	return "custom"
+}
+
+func (cn NetworkCustom) MarshalYAML() (interface{}, error) {
+	return cn.String(), nil
+}
+
+func (cn NetworkCustom) MarshalJSON() ([]byte, error) {
+	return json.RawMessage(`"` + cn.String() + `"`), nil
+}
+
+func (cn NetworkCustom) Set(s string) error {
+	return errors.New("custom networks cannot be set")
+}
+
+func (cn NetworkCustom) Type() string {
+	return "NetworkCustom"
+}
+
+func (cn NetworkCustom) UnmarshalText(text []byte) error {
+	return cn.Set(string(text))
+}
+
+func (cn NetworkCustom) FeederURL() string {
+	return cn.FeederURLVal
+}
+
+func (cn NetworkCustom) GatewayURL() string {
+	return cn.GatewayURLVal
+}
+
+func (cn NetworkCustom) ChainIDString() string {
+	return cn.ChainIDVal
+}
+
+func (cn NetworkCustom) DefaultL1ChainID() *big.Int {
+	return cn.L1ChainIDVal
+}
+
+func (cn NetworkCustom) CoreContractAddress() (common.Address, error) {
+	if cn.CoreContractAddressVal == nil {
+		return common.Address{}, errors.New("core contract address is nil")
+	}
+	return *cn.CoreContractAddressVal, nil
+}
+
+func (cn NetworkCustom) ChainID() *felt.Felt {
+	return new(felt.Felt).SetBytes([]byte(cn.ChainIDVal))
+}
+
+func (cn NetworkCustom) ProtocolID() protocol.ID {
```

Comment:
arent these supposed to be pointer receivers, or am I missing something?

Reply:
For some reason (not sure why), these methods need to be value receivers? Otherwise they don't implement the Network interface

Reply:
We are calling these methods as if they are value receivers (eg as map values, see [line 60 in network_test.go](https://github.com/NethermindEth/juno/blob/main/utils/network_test.go#L60)). 

The codebase could be updated so that we can redefine these methods as pointer receivers, but I expect it will have a non-trivial codebase. I'm happy to leave it as is (a bit biased tho because this is more work for me :) ), or just pushing it to a separate issue.